### DISCUSSION TITLE
Model Predictive Control with autopilot mode which enables the stability of Crazy-flie

### DISCUSSION THREAD

**Discussion by WiresharkIO** [bitcraze/discussions#528] [Q&A] [answered]
I have a functionality based on Model Predictive control(MPC) which determines the position, linear velocities vx, vy, vz and heading angular velocity vphi. 
I want to operate crazyflie in autopilot that controls its stability while I give the velocity commands by the MPC generated values. 

How can i achieve this hover frame in crazyflie that is the autopilot mode which provides me stability with the drone? 

You can also have a look at my repository @ https://github.com/WiresharkIO/Flight-Control/blob/main/motion-with-mpc/MPC-reformulation-Obstacle.py

**Comment by hmllr** [comment, marked as answer]
Hi :)
I think what you want to do is to set a velocity setpoint (an example is here: https://github.com/bitcraze/crazyflie-firmware/blob/b8d2e0b9e9838c548a0a39d2fb3fc5be6b934b73/examples/demos/app_push_demo/src/push.c#L47). This injects the setpoint in the velocity part of the cascade PID controller (as described here: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/controllers/#cascaded-pid-controller), meaning the attitude/rate parts are executed to achieve those setpoints - it should provide you stability, but it is your responsibility to not set your setpoints too extreme.

**Reply by WiresharkIO** [reply]
Thank you @hmllr , I'll try this and get back to you with my observations.

**Comment by arthursn138** [comment]
@WiresharkIO have you ended up implementing it?

**Reply by WiresharkIO** [reply]
@arthursn138 yeah, you can have a look at it here.
https://github.com/WiresharkIO/Flight-Control/blob/03799da831e85527f8558c8e33aa6bcf45bfe868/MPC_Integration/MPC_reformulation_Obstacle.py#L233
